pkg/eestream/encryption.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package eestream
+
+import (
+	"github.com/zeebo/errs"
+)
+
+// Cipher is a type used to define the type of encryption to use
+type Cipher byte
+
+// Constant definitions for no encryption (0), AESGCM (1), and SecretBox (2)
+const (
+	None = Cipher(iota)
+	AESGCM
+	SecretBox
+)
+
+// Constant definitions for key and nonce sizes
+const (
+	GenericKeySize   = 32
+	GenericNonceSize = 24
+	AESGCMNonceSize  = 12
+)
+
+// GenericKey represents the largest key used by any encryption protocol
+type GenericKey [GenericKeySize]byte
+
+// Bytes returns the key as a byte array pointer
+func (key *GenericKey) Bytes() *[GenericKeySize]byte {
+	return (*[GenericKeySize]byte)(key)
+}
+
+// GenericNonce represents the largest nonce used by any encryption protocol
+type GenericNonce [GenericNonceSize]byte
+
+// Bytes returns the nonce as a byte array pointer
+func (nonce *GenericNonce) Bytes() *[GenericNonceSize]byte {
+	return (*[GenericNonceSize]byte)(nonce)
+}
+
+// Increment increments the nonce with the given amount
+func (nonce *GenericNonce) Increment(amount int64) (truncated bool, err error) {
+	return incrementBytes(nonce.Bytes()[:], amount)
+}
+
+// AESGCMNonce returns the nonce as a AES-GCM nonce
+func (nonce *GenericNonce) AESGCMNonce() *AESGCMNonce {
+	aes := new(AESGCMNonce)
+	copy((*aes)[:], (*nonce)[AESGCMNonceSize:])
```

Comment:
Shouldn't this be `(*nonce)[:AESGCMNonceSize]`, similarly everywhere else?

Also this can be simplified to:
```
var aes AESGCMNonce
copy(aes[:], *nonce) // or copy(aes[:], (*nonce)[AESGCMNonceSize:]), whichever is the correct one
return &aes
```

Reply:
Defining `aes` with `var aes AESGCMNonce` without saying `aes = new(AESGCMNonce)` results in a segfault.